server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -69,17 +71,11 @@
         for (FieldAliasMapper fieldAliasMapper : fieldAliasMappers) {
             String aliasName = fieldAliasMapper.name();
             String path = fieldAliasMapper.path();
-            MappedFieldType fieldType = fullNameToFieldType.get(path);
-            fullNameToFieldType.put(aliasName, fieldType);
-            if (fieldType instanceof DynamicFieldType) {
-                dynamicFieldTypes.put(aliasName, (DynamicFieldType) fieldType);
-            }
+            this.runtimeFields.put(aliasName, new AliasRuntimeField(aliasName, path));
         }
 
         for (RuntimeField runtimeField : runtimeFields) {
-            MappedFieldType runtimeFieldType = runtimeField.asMappedFieldType();
-            //this will override concrete fields with runtime fields that have the same name
-            fullNameToFieldType.put(runtimeFieldType.name(), runtimeFieldType);
+            this.runtimeFields.put(runtimeField.name(), runtimeField);
```

Comment:
I am a bit confused about why we need to introduce the distinction between ordinary fields and runtime fields in here. An alias can alias any field, what difference does it make if it's a runtime field or not?

Reply:
It's because `asMappedFieldType()` takes a `MappingLookup` now, so we can't resolve this at construction time.